### PR TITLE
RATIS-881. Fix Failed unit test because test before MiniRaftCluster ready

### PR DESCRIPTION
**Pull request by runzhiwang**
**What's the probelm ?**
image

**What's the reason ?**
The root cause is [RaftServerMetrics::getPeerCommitIndexGauge](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerMetrics.java#L141) happens before [RaftServerMetrics::addPeerCommitIndexGauge](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerMetrics.java#L122).  
When some RaftServerImpl [setRole(RaftPeerRole.LEADER, "changeToLeader")](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java#L345), the statement [waitForLeader](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/test/java/org/apache/ratis/RaftBasicTests.java#L446) succ to get leader and test begin, but [role.startLeaderState](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java#L349) ->
 [new LeaderState](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/RoleInfo.java#L94) ->
[LeaderState::addSenders](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderState.java#L409)->[RaftServerMetrics::addFollower](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerMetrics.java#L106) -> [RaftServerMetrics::addPeerCommitIndexGauge](https://github.com/apache/incubator-ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerMetrics.java#L122) has not finished.

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 1]
```diff
@@ -460,6 +460,20 @@ public String toString() {
...
+    if (!isLeader()) {
+      return false;
+    }
+
+    final LeaderState leaderState = role.getLeaderState().orElse(null);
+    if (leaderState == null || !leaderState.isReady()) {
+      return false;
+    }
+
+    return true;
```
Comment (lokeshj1703) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:474]: We can simplify this as.
`return leaderState != null && leaderState.isReady();`

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 2]
```diff
@@ -460,6 +460,20 @@ public String toString() {
     return null;
   }
 
+  @VisibleForTesting
+  public boolean isLeaderReady() {
+    if (!isLeader()) {
```
Comment (lokeshj1703) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:465]: We can remove this check as leaderState == null covers the case and use this function in checkLeaderState line 451 as well.

**Comment by runzhiwang** [comment]
@lokeshj1703 Thanks for your review. I have updated the patch.

**Comment by lokeshj1703** [comment]
@runzhiwang Thanks for the contribution! @bshashikant Thanks for the review! I have committed the PR to master branch. 
Made a minor change in the commit: Removed @VisibleForTesting annotation on RaftServerImpl@isLeaderReady.

**Pull request closed without merging**